spring-cloud-stream/src/main/java/org/springframework/cloud/stream/function/FunctionConfiguration.java

```diff
@@ -99,83 +101,87 @@
+				Publisher<Object> beginPublishingTrigger = this.setupBindingTrigger(context);
+
+				RootBeanDefinition bd = (RootBeanDefinition) context.getBeanDefinition(functionProperties.getParsedDefinition()[0]);
+				Method factoryMethod = bd.getResolvedFactoryMethod();
+				if (factoryMethod == null) {
+					Object source = bd.getSource();
+					if (source instanceof MethodMetadata) {
+						Class<?> factory = ClassUtils.resolveClassName(((MethodMetadata) source).getDeclaringClassName(), null);
+						Class<?>[] params = FunctionContextUtils.getParamTypesFromBeanDefinitionFactory(factory, bd);
+						factoryMethod = ReflectionUtils.findMethod(factory, ((MethodMetadata) source).getMethodName(), params);
 					}
 				}
-			});
-
-			RootBeanDefinition bd = (RootBeanDefinition) context.getBeanDefinition(functionProperties.getParsedDefinition()[0]);
-			Method factoryMethod = bd.getResolvedFactoryMethod();
-			if (factoryMethod == null) {
-				Object source = bd.getSource();
-				if (source instanceof MethodMetadata) {
-					Class<?> factory = ClassUtils.resolveClassName(((MethodMetadata) source).getDeclaringClassName(), null);
-					Class<?>[] params = FunctionContextUtils.getParamTypesFromBeanDefinitionFactory(factory, bd);
-					factoryMethod = ReflectionUtils.findMethod(factory, ((MethodMetadata) source).getMethodName(), params);
+				Assert.notNull(factoryMethod, "Failed to introspect factory method since it was not discovered for function '"
+								+ functionProperties.getDefinition() + "'");
+				PollableSupplier pollable = factoryMethod.getReturnType().isAssignableFrom(Supplier.class)
+						? AnnotationUtils.findAnnotation(factoryMethod, PollableSupplier.class)
+								: null;
+
+				if (!functionProperties.isComposeFrom() && !functionProperties.isComposeTo()) {
+					integrationFlow = this.integrationFlowFromProvidedSupplier(functionWrapper, functionInspector, beginPublishingTrigger, pollable)
+							.channel("output").get();
 				}
 			}
-			Assert.notNull(factoryMethod, "Failed to introspect factory method since it was not discovered for function '"
-							+ functionProperties.getDefinition() + "'");
-			PollableSupplier pollable = factoryMethod.getReturnType().isAssignableFrom(Supplier.class)
-					? AnnotationUtils.findAnnotation(factoryMethod, PollableSupplier.class)
-							: null;
-
-			if (!functionProperties.isComposeFrom() && !functionProperties.isComposeTo()) {
-				integrationFlow = this.integrationFlowFromProvidedSupplier(functionWrapper, functionInspector, beginPublishingTrigger, pollable)
-						.channel("output").get();
-			}
 		}
-
 		return integrationFlow;
 	}
 
+	/*
+	 * Creates a publishing trigger to ensure Supplier does not begin publishing until binding is created
+	 */
+	private Publisher<Object> setupBindingTrigger(GenericApplicationContext context) {
+		AtomicReference<MonoSink<Object>> triggerRef = new AtomicReference<>();
+		Publisher<Object> beginPublishingTrigger = Mono.create(emmiter -> {
```

Comment:
Consider to use `MonoProcessor` instead. The `onComplete()` will be called from the event instead of this hacks.